### PR TITLE
Add rake task to sync using AusPayNet API

### PR DESCRIPTION
**Pull request by astley92**
Thanks for maintaining this gem!

I recently saw that Australian Payments Network have taken down the FTP server that was previously used to update the database and bank list and have released an API that can be used to update the bsb db content. So figured I'd make this PR and see if you were interested in merging.

## Why

The current process to update the bsb database file is quite manual and can be made easier, especially given Australian Payments Network have introduced an API to fetch BSB data.

## What

Adds a new rake task `rake bsb:sync_bsb_db` that uses the API to fetch latest BSB data and generate a new DB along with an outline of what changes have been made.

Splits the existing rake task into two. This is to ensure that the bank list task can still be run when required since it isn't updated in this PR and also to keep redundancy in case the API is down or something here breaks. There will still be a manual process to fall back to.

Updates README to reflect the above.

## Discussion

This could be taken a step further again by using this rake task in a GHA that would create a PR if any changes are present. I'd be happy to put that together as well if it is something you'd be interested in?

### CODE REVIEW

File: test/tasks/sync_bsb_db_test.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+describe 'sync_bsb_db rake task' do # rubocop:disable Metrics/BlockLength
+  before do
+    ENV.update('AUSPAYNET_SUB_KEY' => 'something')
+    Rake.application.rake_require('../lib/tasks/sync_bsb_db')
+    Rake::Task['bsb:sync_bsb_db'].reenable
+    File.write('test/tmp/bsb_db.json', File.read('test/fixtures/bsb_db.json'))
+  end
+
+  let(:auspaynet_bsb_client_response) do
```
Comment (astley92) [inline test/tasks/sync_bsb_db_test.rb:15]: This is pretty verbose, but can be nice having input and expected output next to eachother in a test that requires a bit of context to understand. Happy to move to a fixture if preferred

File: test/tasks/sync_bsb_db_test.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        }
+      }
+    )
+  end
+
+  it 'generates the expected bsb_db and changes file' do
+    BSB.stub_consts(DB_FILEPATH: 'test/tmp/bsb_db.json', CHANGES_FILEPATH: 'test/tmp/latest_update.json') do
+      BSB::AusPayNet::Client.stub(:fetch_all_bsbs, auspaynet_bsb_client_response) do
+        Rake::Task['bsb:sync_bsb_db'].invoke
+      end
```
Comment (astley92) [inline test/tasks/sync_bsb_db_test.rb:121]: I barely use minitest. Open to feedback if theres better ways to run this setup / stubbing

File: lib/bsb/aus_pay_net/client.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+          headers: {
+            'Content-Type': 'application/json',
+            'Ocp-Apim-Subscription-Key': subscription_key
+          }
+        ) do |faraday|
+          faraday.response :raise_error
+        end
+
+        response = conn.post('/bsbquery/manual/paths/invoke') do |req|
+          req.body = { outputparam: ' ' * OUTPUT_PARAM_WIDTH }.to_json
```
Comment (astley92) [inline lib/bsb/aus_pay_net/client.rb:30]: API guides to have at least 30 chars here

File: test/fixtures/vcr_cassettes/auspaynet_fetch_all_bsbs.yml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      - '19998471'
+      x-ms-request-id:
+      - australiasoutheast:8d349630-46fe-44ae-9d3f-0dc53071b48e
+      request-context:
+      - appId=cid-v1:62fce8e1-7550-4725-83ec-f11232c5768f
+      date:
+      - Fri, 06 Sep 2024 12:30:08 GMT
+    body:
+      encoding: UTF-8
+      string: '{"outputparam":"Query results returned "}[{"BSBCode":"012-002","BSBName":"ANZ
```
Comment (astley92) [inline test/fixtures/vcr_cassettes/auspaynet_fetch_all_bsbs.yml:76]: Manually amended this, the response is huge and felt overkill tracking the whole thing so cut off ~15k lines

File: README.md [hunk 5]
```diff
@@ -68,18 +68,53 @@ Two data sources are used:
...
 At the moment BSB data is a manual download from the Auspaynet site [here](https://bsb.auspaynet.com.au/).
 
-You will need to download two files, place them in `tmp/`:
+You will need to download the Key to Abbreviations and BSB Number file and place it in `tmp/`:
 - `Reference Documents` button > `Key to Abbreviations and BSB Number` in CSV format.
+
+Run the sync task with the files to complete sync of the latest data:
+
+```sh
+rake bsb:sync_bank_list_['tmp/key to abbreviations and bsb numbers (august 2024).csv']
```
Comment (davelooi) [inline README.md:81]: Minor typo `sync_bank_list`

**Review by davelooi — COMMENTED** [review]
Excellent work. Thank you @astley92.
Just a minor nitpick on the typo in the README. Otherwise, good to go.

File: test/tasks/sync_bsb_db_test.rb (hunk 1 above)
Comment (davelooi) [reply test/tasks/sync_bsb_db_test.rb:15]: No issue. Verbose tests are preferred.

**Review by davelooi — APPROVED** [review]
I've tested the rake task, and updated the bank list with it. It worked flawlessly.

Thank you @astley92 👍

**Comment by astley92** [comment]
> I've tested the rake task, and updated the bank list with it. It worked flawlessly.
> 
> Thank you @astley92 👍

Thanks for fixing those up and merging @davelooi I did plan to come back to this tonight haha beat me to it.

No problems, OOC would you be interested in the suggestion in the PR description on automating this a step further through a Github action that utilises this?

Was thinking an action that runs once daily that uses this new task and makes a PR if any changes are present. It would need access to an API key stored in the repo secrets so would need an admin user to make that available. Then the only thing to do when an update is due would be to review and merge a PR

Let me know, I'm happy to put something together :)

**Comment by davelooi** [comment]
Yes, automating this process is an excellent idea.
Before AusPayNet deprecated their old FTP, we had an automated GHA that fetch and create pull request on schedule.

I have added a GHA secret `AUSPAYNET_SUB_KEY` to this repo.

We don't have to run this daily though. AusPayNet only pushes updates to production once a month. Having it run weekly should be sufficient.

> Changes are made to BSBs throughout a month; however, these will not be promoted to
Production until the first working day of the month. For example, changes made during June
2024 were promoted to Production on Monday 1 July 2024.

**Comment by astley92** [comment]
Nice, I did see that it used to exist so re-instating makes sense.

Cool thanks for adding that one, I'll see how I go time-wise but should be able to get a PR up next week sometime, definitely before the next update anyway :)

Weekly sounds good, I'll have a look at previous release timings probably something like end of the first day of the month and weekly from then on.

**Pull request merged**